### ISSUE TITLE
Source.fromPublisher does not pick up attributes correctly

### ISSUE THREAD

**Issue by squadgazzz** [akka/akka#30076]
Looks like `Source.fromPublisher` buffers some data from the source. I couldn't find any way to limit or disable the buffer size.
```
  val source3 = Source.fromPublisher {
    Source.fromIterator(() => Iterator.from(1))
      .takeWhile(_ < 50)
      .map { i =>
        println(s"internal akka: $i")
        i
      }
      .runWith(Sink.asPublisher(false))
  }.map { i =>
    Thread.sleep(2000)
    println(s"akka: $i")
    i
  }
```
log:
```
internal akka: 1
internal akka: 2
internal akka: 3
internal akka: 4
internal akka: 5
internal akka: 6
internal akka: 7
internal akka: 8
internal akka: 9
akka: 1
akka: 2
akka: 3
.
.
.
.
```

Originally I faced with this issue converting from Monix reactive streams to Akka-streams.

**Comment by johanandren** [comment]
Its controlled by the input buffer attribute. You can read more here: https://doc.akka.io/docs/akka/current/stream/stream-rate.html#internal-buffers-and-their-effect

**Comment by squadgazzz** [comment]
Thanks! It'd be great to add information about this to JavaDoc.

**Comment by squadgazzz** [comment]
Also, I have to set these buffer settings globally. I need these settings for a single streaming service only. I don't want to use a separate actor system for that. I can't use another materialize because akka-grpc generated services accept actor systems only.

**Comment by johanandren** [comment]
I think it isn't in the javadoc because it is not something specific the reactive stream integration but applies anywhere there is an async boundary, so it is more something that has a clear place to document in the API docs.

In general you won't get that much control with an async stream with Akka streams, that you will never get more than one element in flight, any operator can introduce explicit or implicit buffers if there is a benefit for their operation.

You can set the attribute for only a single operator or a stream subgraph to not have to specify it with the actorsystem global defaults.

**Comment by squadgazzz** [comment]
> I think it isn't in the javadoc because it is not something specific the reactive stream integration but applies anywhere there is an async boundary, so it is more something that has a clear place to document in the API docs.
> 
> In general you won't get that much control with an async stream with Akka streams, that you will never get more than one element in flight, any operator can introduce explicit or implicit buffers if there is a benefit for their operation.
> 
> You can set the attribute for only a single operator or a stream subgraph to not have to specify it with the actorsystem global defaults.

I use `Source.fromPublisher` because I have to convert from another type of stream. And `Source.fromPublisher(pub).withAttributes(Attributes.inputBufferSize(1,1))` doesn't work for me.

**Comment by johanandren** [comment]
Ah, I see, there is something funky with input buffer (or all attributes) for `Source.fromPublisher`, if I set the attributes on the whole graph, it correctly picks it up:

```
  Source.fromPublisher {
    Source.fromIterator(() => Iterator.from(1))
      .takeWhile(_ < 50)
      .map { i =>
        println(s"internal akka: $i")
        i
      }
      .runWith(Sink.asPublisher(false))
  }.map { i =>
    Thread.sleep(2000)
    println(s"akka: $i")
    i
  }.addAttributes(Attributes.inputBuffer(1, 1))
  .run()
```

**Comment by squadgazzz** [comment]
Oh, wow. Thanks! `.withAttributes` right after `.fromPublisher `doesn't work, but after `.map` it works correctly.

**Comment by Seetaramayya** [comment]
_I am starting baby steps in akka-streams code base, please bare with my annoying (or dumb) questions_. I would like to pick this, but I need help to proceed

I am not able to understand, is it a feature or a bug? 

At the time of materialization, `GraphStageIsland#takePublisher` takes attributes from input side of the connection ( in other words downstream component attributes). 

I can pass source (or publisher)  attributes to downstream but is it good or not I don't know.

**Fix I am thinking about is**: 
```
override def takePublisher(slot: Int, publisher: Publisher[Any], publisherAttributes: Attributes): Unit = {
    val connection = conn(slot)
    val bufferSize = outOwnerAttributes.mandatoryAttribute[InputBuffer].max
    val boundary = new BatchingActorInputBoundary(bufferSize, shell, publisher, "publisher.in")
```

**PhaseIsland** trait looks like this

```
@DoNotInherit private[akka] trait PhaseIsland[M] {
  ...
  @InternalStableApi
  def takePublisher(slot: Int, publisher: Publisher[Any], publisherAttributes: Attributes = Attributes.Empty): Unit

  ...

}
```

**Comment by johanandren** [comment]
Quite a deep end to get your feet wet in, just as a warning :) 

Isn't the `SourceModulePhase` used in this case since `PublisherSource` is a `SourceModule`? In that case the `takePublisher` override throws so that can't be the right place?

**Comment by jrudolph** [comment]
FTR, I think the original behavior was the intended one and should not have been changed. By definition an InputBuffer is in front of a stream component and not after. A Source has no stream component in front of it so changing the InputBuffer attribute should make no difference. That also explains the behavior of putting the attribute after `map` because it changes the `InputBuffer` going *into* map.

In general, expecting tight buffer control in akka-stream is usually an anti-pattern and should be used with much consideration.

**Comment by johanandren** [comment]
I agree that the change was questionable, so maybe we should indeed revert.

I think I've seen some reports of _other_ usage where tight buffer control has been expected that broke with this change. :/ 

For the record I think what made me accept trying to fix it was that with `Source.fromPublisher` there is a buffer "in front"/inside/between the stream and the publisher making it feel like weird ergonomics that what looks like it would control/set attributes for that operator in fact did not. It was not so much about the actual value of being able to tune that buffer.

**Comment by jrudolph** [comment]
> `Source.fromPublisher` there is a buffer "in front"/inside/between the stream and the publisher making it feel like weird ergonomics

Thinking about it again, I can see how you can see `Source.fromPublisher` as the encapsulation of the async boundary between an (external) publisher and an (akka-stream managed) stream, so it would indeed include the inputBuffer.

On the other hand, we actually do handle those publishers and the necessary async boundary transparently, so that the input boundary (and buffer) in question would only be part of the next element. So, tricky call, indeed...
